cpp/cmake_modules/FindRe2.cmake

```diff
@@ -0,0 +1,69 @@
+##############################################################################
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+##############################################################################
+
+# - Find re2 headers and lib.
+# RE2_HOME hints the location
+# This module defines
+#  RE2_INCLUDE_DIR, directory containing headers
+#  RE2_STATIC_LIB, path to libre2.a
+#  re2 imported static library
+#  RE2_FOUND, whether re2 has been found
+
+set(RE2_SEARCH_HEADER_PATHS ${RE2_HOME}/include)
+
+set(RE2_SEARCH_LIB_PATHS ${RE2_HOME}/lib)
+
+find_path(RE2_INCLUDE_DIR re2/re2.h
+  PATHS ${RE2_SEARCH_HEADER_PATHS}
+        NO_DEFAULT_PATH
+  DOC  "Google's re2 regex header path"
+)
+
+find_library(RE2_LIBS NAMES re2
+  PATHS ${RE2_SEARCH_LIB_PATHS}
+        NO_DEFAULT_PATH
+  DOC   "Google's re2 regex library"
+)
+
+find_library(RE2_STATIC_LIB NAMES libre2${CMAKE_STATIC_LIBRARY_SUFFIX}
```

Comment:
@wesm, I have tested this with MSVC and it was able to find library, if it was manually named "libre2.lib". But expected name of static lib on Windows is the question. Have checked cmake script of re2 lib itself, but it doesn't name static build differently to shared build. To have universal solution, it might make sense to introduce configurable separate variable `RE2_STATIC_LIB_NAME`.